### PR TITLE
sysbuild: use relative path for portable

### PR DESCRIPTION
**Pull request by hakehuang**
in the sysbuild cmake the output domians.yaml is asb path, but if we move the test pacakge to another machine this would be a problem, so change it to rel path.

Also use os.path.join in domains in build_helper.py

- build test:

` ./scripts/twister -p mimxrt1060_evk --prep-artifacts-for-testing --package-artifacts twister_test_only.tar.gz -T tset/boot/test_mcuboot`

- run test:

unzip the twister_test_only.tar.gz
`tar cvf twister_test_only.tar.gz`
run command line:
`${ZEPHYR_BASE}/scripts/twister --device-testing --hardware-map <your local map file> --test-only`

### CODE REVIEW

**Review by nordicjm — CHANGES_REQUESTED** [review]
Moving the build folder to another machine is not a supported operation

**Comment by PerMac** [comment]
> Moving the build folder to another machine is not a supported operation

What do you mean? Can you elaborate? It is quite a needed possibility for testing. Example: a company have access to many build nodes, i.e. servers optimize for building. Such servers can have a basic setup without any hw connected. Those can even be  on-demand cloud servers. Then, an execution stage is performed. It does require a special test node, with specific hw connected. The artifacts (images and configs) are downloaded from the previous stage and twister is called in --test-only mode, where it is not building anything. This can vastly optimize the execution. It is already possible and happening with regular tests. I don't understand why you're saying it is not supported.

**Comment by hakehuang** [comment]
> Moving the build folder to another machine is not a supported operation

with this PR, it is supported now.

**Comment by hakehuang** [comment]
@nashif can you help to check and merge this, this fix a test issue whcih match file line number in test report.

**Comment by nordicjm** [comment]
> @nashif can you help to check and merge this, this fix a test issue whcih match file line number in test report.

No absolutely not, as a contributor to sysbuild I am blocking this PR.

**Comment by hakehuang** [comment]
> No absolutely not, as a contributor to sysbuild I am blocking this PR.

@nordicjm may I know why you block it? this feature is supported as I showed above, and sysbuild need this PR to get alignment.

**Review by tejlmand — CHANGES_REQUESTED** [review]
@hakehuang Thanks for raising awareness on this.

While I do see the usefulness on this, then i'm not yet convinced of the approach proposed in this PR.

What I believe @nordicjm is refering to https://github.com/zephyrproject-rtos/zephyr/pull/64551#pullrequestreview-1703807252 is the fact that CMake by design uses absolute paths and does not support relocating a build folder to another machine.

See more on this topic in the CMake faq here:  https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#why-does-cmake-use-full-paths-or-can-i-copy-my-build-tree

Starting to assume certain parts of the build folder is relocatable is a dangerous assumption, cause what can a user expect to be able to relocate ?

Concerning CMake itself, then there do exists a couple of standardized ways to export or package artifacts from a project in a re-usable way, most relevant in this case would be [CPack](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Packaging%20With%20CPack.html) or [install(EXPORT ...)](https://cmake.org/cmake/help/latest/command/install.html#export)

To me, it appears that twister tries to do similar to CPack (based on the arguments: `-prep-artifacts-for-testing --package-artifacts twister_test_only.tar.gz`) but in a way which is unknown and hidden from the build system.

This means, I as build system maintainer have zero knowledge of which files must be kept in relative path, and that is not good.
I wonder why we are inventing a custom packaging mechanism inside twister ?

Besides the above, my main reason for not approving this PR is the fact that sysbuild in future will support pre-built binaries.
Such binaries could be a blob provided through a Zephyr module which contains specific firmware to flash.
Cases currently handled by embedding the binary into the Zephyr image, but an alternative is to flash such binaries independently, especially for multi-core SOcs.

In such scenarios a relative path inside the build folder is not possible, as the binary could reside outside the build folder, and thus blindly copy the domains.yml would not work.

But a CMake packaging tool would have such knowledge and could 1) include the referenced binary, 2) adjust abs paths to rel paths where needed.

**Comment by tejlmand** [comment]
I don't understand what you're trying to describe in the commit message for https://github.com/zephyrproject-rtos/zephyr/pull/64551/commits/230b2e6f7a1f321e3b48c3bcbc108739c35c5eb8:
```
1. in domains setting, each domain should withih the build/dir
this is required by portal.so update the case with this constrains
2. remove runner testing line No. check
```

could you please rephrase, so that I can compare the commit message to the changes.

**Comment by hakehuang** [comment]
@tejlmand , thanks a lot for your detailed explanations, please see my feedback on your concerns.

> What I believe @nordicjm is refering to [#64551 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/64551#pullrequestreview-1703807252) is the fact that CMake by design uses absolute paths and does not support relocating a build folder to another machine.
In this patch we do not change the CMake principle, actually we modify the testrunners.yaml and testplan when we try to wrapped it. So I do not see risk here. and as sysbuild has to use `west flash` so when we do this change and move the package to anopther machine, we do not use CMake anymore.

> Starting to assume certain parts of the build folder is relocatable is a dangerous assumption, cause what can a user expect to be able to relocate ?
It is warrped by twister scripts, and with this change we can move pacakge to another run machine and run twister with. User can think anything. per my understanding, this is a workaround to CMake defects(even by design). we can somehow document this.

> To me, it appears that twister tries to do similar to CPack (based on the arguments: -prep-artifacts-for-testing --package-artifacts twister_test_only.tar.gz) but in a way which is unknown and hidden from the build system.
when we do this, the following steps is done by `west flash` as documented in sysbuild, so there is no build system involved anymore. 

> This means, I as build system maintainer have zero knowledge of which files must be kept in relative path, and that is not good. I wonder why we are inventing a custom packaging mechanism inside twister ?
Because we need this feature for our testing system to test sysbuild, we inverst on what we need, and we are acting as testing working group to maintain this in twister. Otherwise, there is no way to test the sysbuild with server/run machine mode. Beside CPack is another new package it could be phased out, or changed totally, it just out of twister's control, so if we are able to use a small code to keep it for our usage, I do not see any objection to do this way.

> Besides the above, my main reason for not approving this PR is the fact that sysbuild in future will support pre-built binaries.
Such binaries could be a blob provided through a Zephyr module which contains specific firmware to flash.
Cases currently handled by embedding the binary into the Zephyr image, but an alternative is to flash such binaries independently, especially for multi-core SOcs.
the problem is how far the future will be? we need this feautre now, I can definitly revert it if there is a better solution. and I find you make things complex, seeing from sysbuild it just read the domains.yaml and runners.yaml, and decide what to flahsh from where, this is a pure data driver actions(Thanks to good disgn in sysbuild), if sysbuild keeps the data driven design, why we need to design another mechnism just to update the relative path? The blob firmware path can be varied for exmaple, there is a porprietary version of same fw that need to be tested, a simple step is to just modify the domains.yaml and update the path, no build system can ever cover all requirement, we need have flexibilities.

> But a CMake packaging tool would have such knowledge and could 1) include the referenced binary, 2) adjust abs paths to rel paths where needed.
As I said, if sysbuild keeps data driven way, we can use any mothed to update the data, not only rely on CMake, which is not deigned only for zephyr. if a plugin is required, why just support the data updates in twister like we do here?

**Comment by hakehuang** [comment]
> I don't understand what you're trying to describe in the commit message for [230b2e6](https://github.com/zephyrproject-rtos/zephyr/commit/230b2e6f7a1f321e3b48c3bcbc108739c35c5eb8):
> 
> ```
> 1. in domains setting, each domain should withih the build/dir
> this is required by portal.so update the case with this constrains
> 2. remove runner testing line No. check
> ```
> 
> could you please rephrase, so that I can compare the commit message to the changes.

this is just to make the test compatible with what we changed, I will rephase it.

**Comment by nordicjm** [comment]
> In this patch we do not change the CMake principle, actually we modify the runners.yaml and domains.yaml when we try to wrapped it. So I do not see risk here. And as sysbuild uses west flash to flash, so when we do this change and move the package to anopther machine, we do not use CMake anymore.

This is not true, doing `west flash` will call cmake

**Comment by hakehuang** [comment]
> This is not true, doing `west flash` will call cmake

@nordicjm it only search for a CMakeCache.txt, it will not call cmake

**Comment by nordicjm** [comment]
> > This is not true, doing `west flash` will call cmake
> 
> @nordicjm it only searches for a CMakeCache.txt, it will not call cmake. it is based on Cmake result, but not call CMake again.

So here's what happens when you configure a build and build on one machine:
```
in zephyr/sample/hello_world
mkdir build
cd build
cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -DAPP_DIR=.. -DSB_CONFIG_BOOTLOADER_MCUBOOT=y /tmp/aa/zephyr/share/sysbuild/
ninja
cd ..
tar cf out.tar build
```
Then when you switch to another system:
```
scp blah@blah:blah/out.tar .
tar xf out.tar
cd build
west flash

-- west flash: rebuilding
[0/1] Re-running CMake...
Loading Zephyr module(s) (Zephyr base (cached)): extensions sysbuild_extensions python west root zephyr_module boards shields sysbuild_kconfig
-- Found Python3: /usr/bin/python (found suitable version "3.11.5", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/jamie/.cache/zephyr
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: nrf52840dk_nrf52840
Parsing /tmp/aa/zephyr/share/sysbuild/Kconfig
Loaded configuration '/tmp/aa/zephyr/samples/hello_world/_AA/zephyr/.config'
No change to configuration in '/tmp/aa/zephyr/samples/hello_world/_AA/zephyr/.config'
No change to Kconfig header in '/tmp/aa/zephyr/samples/hello_world/_AA/autoconf.h'
-- 
   *********************************
   * Running CMake for hello_world *
   *********************************

Loading Zephyr default modules (Zephyr base (cached)).
-- Application: /tmp/aa/zephyr/samples/hello_world
```
So yes, it does

**Comment by hakehuang** [comment]
> So yes, it does. The application gets reconfigured and rebuilt.

@nordicjm , there is an error in your last step, we do not call `west falsh`, but call `west flash --skip-rebuild`

File: scripts/pylib/twister/twisterlib/runner.py
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/runner.py:786]: Is this needed? I though we figure out which binaries are required based on runners.yaml https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/runner.py#L784

Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/runner.py:786]: this can be optional, I think this can keep options open when do real flash

Comment (PerMac) [reply scripts/pylib/twister/twisterlib/runner.py:786]: what can be kept optional? runners.yaml defines which files you'll need for flashing. this in not optional. It seems in case of sysbuild runners.yaml has slightly different location than in regular cases, hence it don't find it with the current code. With the path updated for sysbuild, all binaries are properly added, including signed ones.

**Review by PerMac — CHANGES_REQUESTED** [review]

File: scripts/pylib/twister/twisterlib/runner.py
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/runner.py:786]: I see you point, let me update

File: scripts/pylib/twister/twisterlib/runner.py
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/runner.py:959]: ok, let me update

File: scripts/pylib/twister/twisterlib/runner.py
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/runner.py:918]: let me check

**Comment by hakehuang** [comment]
@mbolivar-ampere @PerMac @nordicjm can you revisit this. I have updated according to your comments

**Comment by tejlmand** [comment]
> the problem is how far the future will be?

Until those who need the feature opens a PR.
Already today there are support for binary blobs, and I know there are users who looks at how to have those part of sysbuild.

Currently such blobs may be merged into the zephyr binary.
As there do exists alternative then I don't know the exact prioritization for extending sysbuild for those who need the feature.

> we need this feautre now, we can definitly revert it if there is a better solution. 

I understand your urgent need, and having someone utilizing sysbuild and proposing improvements is surely a very good thing and highly appreciated.
However, I also get concerned with statement like this, because contributors has little incitement to change things if things are working for them, meaning such a change would likely put burden on someone else to further improve the design later.

File: scripts/pylib/twister/twisterlib/runner.py
Comment (tejlmand) [inline scripts/pylib/twister/twisterlib/runner.py:906]: oh my, this seems like `twister` is manipulating the files produced by the build system.

Twister should really not be messing with files created by the build system.
If twister wants to update such files before packaging for relocation purposes, then please do so in a dedicated folder in the build tree, and not by touching files controlled by the build system.

I'm not really keen on the possible risk of having to track down bugs because twister has changed content behind the back of the build system.

I notice this is not something new, but the knowledge about it has been unknown to me until reviewing this PR.

@nordicjm and @PerMac thoughts ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
A couple of change requests.

File: scripts/pylib/twister/twisterlib/runner.py
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/runner.py:906]: @tejlmand , I am very unacceptible on this, the former twister change method is accepted and I can not be accept by doing the sameway. so I think this is just to block me to change this. so I will resort to use other tools, not in sysbuild.

Comment (tejlmand) [reply scripts/pylib/twister/twisterlib/runner.py:906]: excuse me ?

I did not block the PR over this, but when noticing potential issues in code should in my opinion always be highlighted and relevant maintainers be notified.

As I mention in the comment,
> I'm not keen on this

and
> I notice this is not something new

is not the same as comments given here, where the wording `not accepted / approved` is used:
https://github.com/zephyrproject-rtos/zephyr/pull/64551#discussion_r1404264176
https://github.com/zephyrproject-rtos/zephyr/pull/64551#discussion_r1404272056

or do @hakehuang mean that code with potential side effects should not be highlighted / relevant people notified when such code is noticed ?

Comment (tejlmand) [reply scripts/pylib/twister/twisterlib/runner.py:906]: and @hakehuang the comment should at least be addressed, addressing could be in the form:
- Issue will be fixed immediately
- We accept for now, but as there haven't been reports of problems in practice, but will create follow-up enhancement to address in future / make visible to others that a potential problem exists.
- thanks for raising awareness, we accept the risk / annoyance but consider it low enough to ignore
- ....

also remember that in future when someone does a git blame then this PR will show up, and others will be able to find this comment and understand why code is like this / why it was accepted.

Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/runner.py:906]: > and @hakehuang the comment should at least be addressed, addressing could be in the form:
> 
> * Issue will be fixed immediately
> * We accept for now as there haven't been reports of problems in practice, but will create follow-up enhancement to address in future / make visible to others that a potential problem exists.
> * thanks for raising awareness, we accept the risk / annoyance but consider it low enough to ignore
> * ....
> 
> also remember that in future when someone does a git blame then this PR will show up, and others will be able to find this comment and understand why code is like this / why it was accepted.

@tejlmand , you are the code owner, and I do respect your judegement. And as this feature enhancement takes so much time for me. and I have to think about another approach for this.

**Pull request closed without merging**
